src/runtime/sharecommon.cpp

```diff
@@ -13,8 +13,8 @@ namespace lean {
 extern "C" LEAN_EXPORT uint8 lean_sharecommon_eq(b_obj_arg o1, b_obj_arg o2) {
     lean_assert(!lean_is_scalar(o1));
     lean_assert(!lean_is_scalar(o2));
-    size_t sz1 = lean_object_byte_size(o1);
-    size_t sz2 = lean_object_byte_size(o2);
+    size_t sz1 = lean_object_data_size(o1);
+    size_t sz2 = lean_object_data_size(o2);
     if (sz1 != sz2) return false;
```

Comment:
It's possible that this should reject objects as unequal if they have the same 'value' but differing capacities.

Reply:
That _used_ to be the case on account of L18, but is now no longer so. As long as the "value" is the same, we now compare equal regardless of capacity.

Reply:
Right, I was pointing this out for other reviewers; perhaps we should be more conservative and _only_ fix the undefined behavior without changing the behavior for objects with differing numbers of unused bytes.

Reply:
So you'd consider the total _size_ (but not contents) as salient, too? We could do that. It'd be good to then also include that size in the hash.

I'm not sure if this is overall desirable, though. I.e. why would you expect equal objects to have the same capacity? Wouldn't the capacity be checked locally by any operation that grows the contents, and not something you'd assume upfront?

That is, yes, we could make this change, but I'd hope that nothing would depend on that.

Reply:
> we now compare equal regardless of capacity (and also hash equal)

How so, is `m_capacity` not part of the `memcmp` (and hash) as it is located after the `lean_object` header? I think your implementation is fine as is; I *don't* think we can assume that all `ShareCommon` users do not care about preserving capacity, identifying two objects with differing capacity may break referential transparency.